File: examples/ios/VistreaDemoApp/Sources/ScenarioModels.swift
import Foundation

struct ScenarioManifest: Decodable, Sendable {
    let suiteID: String
    let scenarios: [ScenarioManifestEntry]

    private enum CodingKeys: String, CodingKey {
        case suiteID = "suite_id"
        case scenarios
    }
}
struct ScenarioManifestEntry: Decodable, Sendable {
    let scenarioID: String
    let file: String
    let required: Bool

    private enum CodingKeys: String, CodingKey {
        case scenarioID = "scenario_id"
        case file
        case required
    }
}

struct ScenarioDefinition: Decodable, Sendable {
    let scenarioID: String
    let title: String
    let purpose: String
    let profiles: [String]
    let reset: ScenarioReset
    let stableNodes: [ScenarioStableNode]
    let states: [ScenarioState]
    let steps: [ScenarioStep]

    var entryStateID: String {
        reset.entryStateID
    }

    func state(id: String) -> ScenarioState? {
        states.first { $0.stateID == id }
    }

    func stableNode(id: String) -> ScenarioStableNode? {
        stableNodes.first { $0.nodeID == id }
    }

    func steps(from stateID: String, profile: String) -> [ScenarioStep] {
        steps.filter {
            $0.fromStateID == stateID && ($0.profiles?.contains(profile) ?? true)
        }
    }

    private enum CodingKeys: String, CodingKey {
        case scenarioID = "scenario_id"
        case title
        case purpose
        case profiles
        case reset
        case stableNodes = "stable_nodes"
        case states
        case steps
    }
}

struct ScenarioReset: Decodable, Sendable {
    let entryStateID: String
    let seed: Int

    private enum CodingKeys: String, CodingKey {
        case entryStateID = "entry_state_id"
        case seed
    }
}

struct ScenarioStableNode: Decodable, Sendable {
    let nodeID: String
    let role: String

    private enum CodingKeys: String, CodingKey {
        case nodeID = "node_id"
        case role
    }
}

struct ScenarioState: Decodable, Sendable {
    let stateID: String
    let kind: String
    let requiredNodeIDs: [String]

    private enum CodingKeys: String, CodingKey {
        case stateID = "state_id"
        case kind
        case requiredNodeIDs = "required_node_ids"
    }
}

struct ScenarioStep: Decodable, Sendable {
    let stepID: String
    let action: ScenarioAction
    let fromStateID: String
    let toStateID: String
    let profiles: [String]?

    private enum CodingKeys: String, CodingKey {
        case stepID = "step_id"
        case action
        case fromStateID = "from_state_id"
        case toStateID = "to_state_id"
        case profiles
    }
}

struct ScenarioAction: Decodable, Sendable {
    let kind: String
    let targetNodeID: String?
    let inputAlias: String?
    let durationMilliseconds: Int?

    private enum CodingKeys: String, CodingKey {
        case kind
        case targetNodeID = "target_node_id"
        case inputAlias = "input_alias"
        case durationMilliseconds = "duration_ms"
    }
}

File: examples/android/VistreaDemoApp/app/src/main/kotlin/dev/vistrea/demo/contract/ScenarioModels.kt
package dev.vistrea.demo.contract

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject

@Serializable
data class ScenarioManifest(
    val schema_version: Int,
    val suite_id: String,
    val protocol_version: JsonObject,
    val required_platforms: List<String>,
    val platforms: JsonObject,
    val profiles: List<LaunchProfile>,
    val scenarios: List<ScenarioManifestEntry>,
    val vertical_loops: JsonArray,
    val extensions: JsonObject,
)

@Serializable
data class LaunchProfile(
    val profile_id: String,
    val base_profile: String? = null,
    val description: String,
    val determinism: JsonObject,
)

@Serializable
data class ScenarioManifestEntry(
    val scenario_id: String,
    val file: String,
    val required: Boolean,
    val coverage: List<String>,
)

@Serializable
data class ScenarioFixture(
    val schema_version: Int,
    val scenario_id: String,
    val title: String,
    val purpose: String,
    val required: Boolean,
    val coverage: List<String>,
    val platform_support: JsonObject,
    val profiles: List<String>,
    val reset: ScenarioReset,
    val stable_nodes: List<StableNode>,
    val states: List<ScenarioState>,
    val steps: List<ScenarioStep>,
    val expected_artifacts: JsonArray,
    val expectations: JsonObject,
    val extensions: JsonObject,
)

@Serializable
data class ScenarioReset(
    val entry_state_id: String,
    val seed: Int,
    val external_services: Boolean,
)

@Serializable
data class StableNode(
    val node_id: String,
    val role: String,
)

@Serializable
data class ScenarioState(
    val state_id: String,
    val kind: String,
    val required_node_ids: List<String>,
    val dynamic_fields: JsonArray,
)

@Serializable
data class ScenarioStep(
    val step_id: String,
    val action: ScenarioAction,
    val from_state_id: String,
    val to_state_id: String,
    val profiles: List<String> = emptyList(),
)

@Serializable
data class ScenarioAction(
    val kind: String,
    val target_node_id: String? = null,
    val input_alias: String? = null,
    val duration_ms: Int? = null,
)

data class ScenarioSuite(
    val manifest: ScenarioManifest,
    val fixturesById: Map<String, ScenarioFixture>,
) {
    fun fixture(scenarioId: String): ScenarioFixture =
        requireNotNull(fixturesById[scenarioId]) { "Unknown Scenario ID: $scenarioId" }
}

data class StableNodeIdentity(
    val contentDescription: String,
    val tag: String,
) {
    companion object {
        fun from(nodeId: String): StableNodeIdentity {
            require(nodeId.startsWith("demo.")) { "Shared stable node IDs must use demo.*." }
            return StableNodeIdentity(contentDescription = nodeId, tag = nodeId)
        }
    }
}
